### DISCUSSION TITLE
Using LittleFS with ESP32 but app still compiles with SPIFFS+LittleFS

### DISCUSSION THREAD

**Discussion by HankLloydRight** [Hieromon/AutoConnect#471] [General]
Hello,
 I've converted my main ESP32 app to use LittleFS instead of SPIFFS, which is working great. But I still see SPIFFS being compiled in the dependency tree due to AutoConnect and Pagebuilder: 

```
-- <LittleFS_esp32> 1.0.6
|   |-- <FS> 1.0
|-- <AutoConnect> 1.3.3
|   |-- <PageBuilder> 1.5.2
|   |   |-- <FS> 1.0
|   |   |-- <LittleFS_esp32> 1.0.6
|   |   |   |-- <FS> 1.0
|   |   |-- <SPIFFS> 1.0
|   |   |   |-- <FS> 1.0
|   |   |-- <WebServer> 1.0
|   |   |   |-- <WiFi> 1.0
|   |   |   |-- <FS> 1.0
|   |   |-- <WiFi> 1.0
|   |-- <ArduinoJson> 6.19.2
|   |-- <DNSServer> 1.1.0
|   |   |-- <WiFi> 1.0
|   |-- <EEPROM> 1.0.3
|   |-- <WebServer> 1.0
|   |   |-- <WiFi> 1.0
|   |   |-- <FS> 1.0
|   |-- <WiFi> 1.0
|   |-- <Preferences> 1.0
|   |-- <LittleFS_esp32> 1.0.6
|   |   |-- <FS> 1.0
|   |-- <FS> 1.0
|   |-- <SPIFFS> 1.0
|   |   |-- <FS> 1.0
|   |-- <Update> 1.0
|   |-- <Ticker> 1.1
|   |-- <HTTPClient> 1.2
|   |   |-- <WiFi> 1.0
|   |   |-- <WiFiClientSecure> 1.0
|   |   |   |-- <WiFi> 1.0
|   |-- <HTTPUpdate> 1.3
|   |   |-- <HTTPClient> 1.2
|   |   |   |-- <WiFi> 1.0
|   |   |   |-- <WiFiClientSecure> 1.0
|   |   |   |   |-- <WiFi> 1.0
|   |   |-- <Update> 1.0
|   |   |-- <WiFi> 1.0
|   |-- <SD(esp32)> 1.0.5
|   |   |-- <FS> 1.0
|   |   |-- <SPI> 1.0
|   |-- <SPI> 1.0
|-- <SD_MMC> 1.0
|   |-- <FS> 1.0
|-- <NimBLE-Arduino> 1.3.7
|-- <Preferences> 1.0
|-- <ESP8266 and ESP32 OLED driver for SSD1306 displays> 4.2.1
|   |-- <SPI> 1.0
|   |-- <Wire> 1.0.1
```

The code compiles clean and runs great, even with both SPIFFS and LITTLEFS in the tree. I'd like to figure out how to not have SPIFFS compiled into the code, hopefully freeing up some code and stack memory.   For the record, I am not using any of the Autoconnect features of uploading files or custom web pages that actually need to write to the local FS.  

I've read this entire page: https://hieromon.github.io/AutoConnect/filesystem.html#selecting-appropriate-filesystem and I've added this to my platformio.ini file:

```
board_build.filesystem = littlefs   ; was already there
lib_ldf_mode=deep
build_flags =
    -DAC_USE_LITTLEFS
    -DPB_USE_LITTLEFS
```

But when I do that, I get a long list of compile errors (see below).   I've search and search and tried everything I could find, but can't figure out how to get this to compile with only LITTLEFS.  I'm sure it's probably one little thing that's triggering all these errors, but I sure can't figure it out.  Help me Hieromon, you're my only hope! 

Thank you, and thank you for your awesome library! 

-HLR

```
.pio\libdeps\lolin32\AutoConnect\src/AutoConnectFS.h:60:47: error: 'LittleFSFS' in namespace 'fs' does not name a type
 #define AUTOCONNECT_APPLIED_FILECLASS     fs::LittleFSFS

.pio\libdeps\lolin32\AutoConnect\src/AutoConnectFS.h:66:14: note: in expansion of macro 'AUTOCONNECT_APPLIED_FILECLASS'
   using FS = AUTOCONNECT_APPLIED_FILECLASS;

.pio\libdeps\lolin32\AutoConnect\src/AutoConnectFS.h:67:41: error: 'AutoConnectFS::_isMounted' declared as an 'inline' variable
   inline bool _isMounted(AutoConnectFS::FS* filesystem) {

.pio\libdeps\lolin32\AutoConnect\src/AutoConnectFS.h:67:26: error: 'FS' is not a member of 'AutoConnectFS'
   inline bool _isMounted(AutoConnectFS::FS* filesystem) {
                          ^
.pio\libdeps\lolin32\AutoConnect\src/AutoConnectFS.h:67:26: note: suggested alternatives:
In file included from .pio\libdeps\lolin32\LittleFS_esp32\src/LITTLEFS.h:17:0,
                 from src\main.h:4,
                 from src\ILDAC-WIFI.cpp:2:
C:\.platformio\packages\framework-arduinoespressif32\libraries\FS\src/FS.h:86:7: note:   'fs::FS'
 class FS
       ^
C:\.platformio\packages\framework-arduinoespressif32\libraries\FS\src/FS.h:86:7: note:   'fs::FS'
In file included from .pio\libdeps\lolin32\AutoConnect\src/AutoConnectOTA.h:25:0,
                 from .pio\libdeps\lolin32\AutoConnect\src/AutoConnect.h:39,
                 from src\main.h:40,
.pio\libdeps\lolin32\AutoConnect\src/AutoConnectFS.h:67:45: error: 'filesystem' was not declared in this scope
   inline bool _isMounted(AutoConnectFS::FS* filesystem) {
.pio\libdeps\lolin32\AutoConnect\src/AutoConnectFS.h:67:57: error: expected ',' or ';' before '{' token
   inline bool _isMounted(AutoConnectFS::FS* filesystem) {
In file included from .pio\libdeps\lolin32\AutoConnect\src/AutoConnect.h:39:0,
                 from src\main.h:40,
                 from src\ILDAC-WIFI.cpp:2:
.pio\libdeps\lolin32\AutoConnect\src/AutoConnectOTA.h:99:18: error: 'FS' in namespace 'AutoConnectFS' does not name a type
   AutoConnectFS::FS*  _fs;      /**< Filesystem for the native file uploading */
                  ^
In file included from .pio\libdeps\lolin32\AutoConnect\src/AutoConnectOTA.h:25:0,
                 from .pio\libdeps\lolin32\AutoConnect\src/AutoConnect.h:39,
                 from src\main.h:40,
                 from src\ILDAC-ILDA.cpp:1:
.pio\libdeps\lolin32\AutoConnect\src/AutoConnectFS.h:60:47: error: 'LittleFSFS' in namespace 'fs' does not name a type
 #define AUTOCONNECT_APPLIED_FILECLASS     fs::LittleFSFS
                                               ^
.pio\libdeps\lolin32\AutoConnect\src/AutoConnectFS.h:66:14: note: in expansion of macro 'AUTOCONNECT_APPLIED_FILECLASS'
   using FS = AUTOCONNECT_APPLIED_FILECLASS;
              ^
.pio\libdeps\lolin32\AutoConnect\src/AutoConnectFS.h:67:41: error: 'AutoConnectFS::_isMounted' declared as an 'inline' variable
   inline bool _isMounted(AutoConnectFS::FS* filesystem) {
                                         ^
.pio\libdeps\lolin32\AutoConnect\src/AutoConnectFS.h:67:26: error: 'FS' is not a member of 'AutoConnectFS'
   inline bool _isMounted(AutoConnectFS::FS* filesystem) {
                          ^
.pio\libdeps\lolin32\AutoConnect\src/AutoConnectFS.h:67:26: note: suggested alternatives:
In file included from .pio\libdeps\lolin32\LittleFS_esp32\src/LITTLEFS.h:17:0,
                 from src\main.h:4,
                 from src\ILDAC-ILDA.cpp:1:
C:\.platformio\packages\framework-arduinoespressif32\libraries\FS\src/FS.h:86:7: note:   'fs::FS'
 class FS
       ^
C:\.platformio\packages\framework-arduinoespressif32\libraries\FS\src/FS.h:86:7: note:   'fs::FS'
In file included from .pio\libdeps\lolin32\AutoConnect\src/AutoConnectOTA.h:25:0,
                 from .pio\libdeps\lolin32\AutoConnect\src/AutoConnect.h:39,
                 from src\main.h:40,
                 from src\ILDAC-ILDA.cpp:1:
.pio\libdeps\lolin32\AutoConnect\src/AutoConnectFS.h:67:45: error: 'filesystem' was not declared in this scope
   inline bool _isMounted(AutoConnectFS::FS* filesystem) {
                                             ^
In file included from .pio\libdeps\lolin32\AutoConnect\src/AutoConnectOTA.h:25:0,
                 from .pio\libdeps\lolin32\AutoConnect\src/AutoConnect.h:39,
                 from src\main.h:40,
                 from src\ILDAC-BLE.cpp:2:
.pio\libdeps\lolin32\AutoConnect\src/AutoConnectFS.h:60:47: error: 'LittleFSFS' in namespace 'fs' does not name a type
 #define AUTOCONNECT_APPLIED_FILECLASS     fs::LittleFSFS
                                               ^
.pio\libdeps\lolin32\AutoConnect\src/AutoConnectFS.h:66:14: note: in expansion of macro 'AUTOCONNECT_APPLIED_FILECLASS'
   using FS = AUTOCONNECT_APPLIED_FILECLASS;
              ^
.pio\libdeps\lolin32\AutoConnect\src/AutoConnectFS.h:67:41: error: 'AutoConnectFS::_isMounted' declared as an 'inline' variable
   inline bool _isMounted(AutoConnectFS::FS* filesystem) {
                                         ^
.pio\libdeps\lolin32\AutoConnect\src/AutoConnectFS.h:67:26: error: 'FS' is not a member of 'AutoConnectFS'
   inline bool _isMounted(AutoConnectFS::FS* filesystem) {
                          ^
.pio\libdeps\lolin32\AutoConnect\src/AutoConnectFS.h:67:26: note: suggested alternatives:
.pio\libdeps\lolin32\AutoConnect\src/AutoConnectFS.h:67:57: error: expected ',' or ';' before '{' token
   inline bool _isMounted(AutoConnectFS::FS* filesystem) {
                                                         ^
In file included from .pio\libdeps\lolin32\LittleFS_esp32\src/LITTLEFS.h:17:0,
                 from src\main.h:4,
                 from src\ILDAC-BLE.cpp:2:
C:\.platformio\packages\framework-arduinoespressif32\libraries\FS\src/FS.h:86:7: note:   'fs::FS'
 class FS
       ^
C:\.platformio\packages\framework-arduinoespressif32\libraries\FS\src/FS.h:86:7: note:   'fs::FS'
In file included from .pio\libdeps\lolin32\AutoConnect\src/AutoConnectOTA.h:25:0,
                 from .pio\libdeps\lolin32\AutoConnect\src/AutoConnect.h:39,
                 from src\main.h:40,
                 from src\ILDAC-BLE.cpp:2:
.pio\libdeps\lolin32\AutoConnect\src/AutoConnectFS.h:67:45: error: 'filesystem' was not declared in this scope
   inline bool _isMounted(AutoConnectFS::FS* filesystem) {
                                             ^
.pio\libdeps\lolin32\AutoConnect\src/AutoConnectFS.h:67:57: error: expected ',' or ';' before '{' token
   inline bool _isMounted(AutoConnectFS::FS* filesystem) {
                                                         ^
In file included from .pio\libdeps\lolin32\AutoConnect\src/AutoConnect.h:39:0,
                 from src\main.h:40,
                 from src\ILDAC-BLE.cpp:2:
.pio\libdeps\lolin32\AutoConnect\src/AutoConnectOTA.h:99:18: error: 'FS' in namespace 'AutoConnectFS' does not name a type
   AutoConnectFS::FS*  _fs;      /**< Filesystem for the native file uploading */
                  ^
In file included from .pio\libdeps\lolin32\AutoConnect\src/AutoConnect.h:39:0,
                 from src\main.h:40,
                 from src\ILDAC-ILDA.cpp:1:
.pio\libdeps\lolin32\AutoConnect\src/AutoConnectOTA.h:99:18: error: 'FS' in namespace 'AutoConnectFS' does not name a type
   AutoConnectFS::FS*  _fs;      /**< Filesystem for the native file uploading */
                  ^
In file included from src\main.h:11:0,
                 from src\ILDAC-WIFI.cpp:2:
src\icons.h:27:16: warning: 'bluetooth_bits' defined but not used [-Wunused-variable]
 static uint8_t bluetooth_bits[] = {
                ^
In file included from .pio\libdeps\lolin32\AutoConnect\src/AutoConnectOTA.h:25:0,
                 from .pio\libdeps\lolin32\AutoConnect\src/AutoConnect.h:39,
                 from src\main.h:40,
                 from src\main.cpp:5:
.pio\libdeps\lolin32\AutoConnect\src/AutoConnectFS.h:60:47: error: 'LittleFSFS' in namespace 'fs' does not name a type
 #define AUTOCONNECT_APPLIED_FILECLASS     fs::LittleFSFS
                                               ^
.pio\libdeps\lolin32\AutoConnect\src/AutoConnectFS.h:66:14: note: in expansion of macro 'AUTOCONNECT_APPLIED_FILECLASS'
   using FS = AUTOCONNECT_APPLIED_FILECLASS;
              ^
.pio\libdeps\lolin32\AutoConnect\src/AutoConnectFS.h:67:41: error: 'AutoConnectFS::_isMounted' declared as an 'inline' variable
   inline bool _isMounted(AutoConnectFS::FS* filesystem) {
                                         ^
.pio\libdeps\lolin32\AutoConnect\src/AutoConnectFS.h:67:26: error: 'FS' is not a member of 'AutoConnectFS'
   inline bool _isMounted(AutoConnectFS::FS* filesystem) {
                          ^
.pio\libdeps\lolin32\AutoConnect\src/AutoConnectFS.h:67:26: note: suggested alternatives:
In file included from .pio\libdeps\lolin32\LittleFS_esp32\src/LITTLEFS.h:17:0,
                 from src\main.h:4,
                 from src\main.cpp:5:
C:\.platformio\packages\framework-arduinoespressif32\libraries\FS\src/FS.h:86:7: note:   'fs::FS'
 class FS
       ^
C:\.platformio\packages\framework-arduinoespressif32\libraries\FS\src/FS.h:86:7: note:   'fs::FS'
In file included from .pio\libdeps\lolin32\AutoConnect\src/AutoConnectOTA.h:25:0,
                 from .pio\libdeps\lolin32\AutoConnect\src/AutoConnect.h:39,
                 from src\main.h:40,
                 from src\main.cpp:5:
.pio\libdeps\lolin32\AutoConnect\src/AutoConnectFS.h:67:45: error: 'filesystem' was not declared in this scope
   inline bool _isMounted(AutoConnectFS::FS* filesystem) {
                                             ^
.pio\libdeps\lolin32\AutoConnect\src/AutoConnectFS.h:67:57: error: expected ',' or ';' before '{' token
   inline bool _isMounted(AutoConnectFS::FS* filesystem) {
                                                         ^
In file included from src\main.h:11:0,
                 from src\ILDAC-BLE.cpp:2:

In file included from .pio\libdeps\lolin32\AutoConnect\src/AutoConnect.h:39:0,
                 from src\main.h:40,
                 from src\main.cpp:5:
.pio\libdeps\lolin32\AutoConnect\src/AutoConnectOTA.h:99:18: error: 'FS' in namespace 'AutoConnectFS' does not name a type
   AutoConnectFS::FS*  _fs;      /**< Filesystem for the native file uploading */
                  ^

```

**Comment by Hieromon** [comment]
@HankLloydRight The problem you are encountering is due to the difference between the version of the [LittleFS library that had integrated into the ESP32 core 2.0](https://github.com/espressif/arduino-esp32/tree/master/libraries/LittleFS) and an ordinary version published on the [ESP32 LittleFS repository](https://github.com/lorol/LITTLEFS).

The original *LittleFS_esp32* deploys Little FS class and instance as **fs::LITTLEFSFS** and **LITTLEFS**: https://github.com/lorol/LITTLEFS/blob/f0817ca5264745acce697092de2bf218b3aa5b2e/src/LITTLEFS.h#L39

On the other hand, in ESP32 Core 2.0, they are declared as **fs::LittleFSFS** and **LittleFS**: https://github.com/espressif/arduino-esp32/blob/c2e5957f35b0441e7c954e247437456fabe3800f/libraries/LittleFS/src/LittleFS.h#L39

This change in the ESP32 core 2.0 was made to ensure compatibility with the ESP8266, but as a result, backward compatibility with legacy code for [LittleFS_esp32](https://github.com/lorol/LITTLEFS) has been lost. This problem does not occur when building in the Arduino IDE environment because it assumes that LittleFS is integrated with the ESP32 2.0 core. And its class and instance are `fs::LittleFSFS` and `LittleFS`, which are consistent with AutoConnect and PageBuilder.

I temporarily implemented a way to absorb the above gap in the AutoConnect library to support ESP32 LittleFS in AutoConnect and PageBuilder, but eventually dropped it. That's because we estimated that it would not take much time for PlatformIO to support ESP32 core 2.0.
But now, my prediction was off. It looks like I need to support LittleFS (it's legacy) in ESP32 core 1.0.6. So I'll forward this topic to Issue for other users as well.

### How to avoid the problem for now:

You will need to employ one of the following two methods. The following workarounds have their limitations. Those limitations will be lifted in the next release as AutoConnect v1.3.4 and PageBuilder v 1.5.3.

1. Modify the [`AutoConnectFS.h`](https://github.com/Hieromon/AutoConnect/blob/master/src/AutoConnectFS.h) and [`PageBuilder.h`](https://github.com/Hieromon/PageBuilder/blob/master/src/PageBuilder.h)

    If you apply this fix, you will not be able to build sketches with the ESP32 Core 2.0 component or ESP8266 Core applied in the Arudino IDE. This will cause a compile error.

    ---
    https://github.com/Hieromon/AutoConnect/blob/4bef741ab82665e0da87333dd02ea701668f7d45/src/AutoConnectFS.h#L55

    Change to:
    ```cpp
    #define AUTOCONNECT_APPLIED_FILESYSTEM    LITTLEFS
    ```

    ---
    https://github.com/Hieromon/AutoConnect/blob/4bef741ab82665e0da87333dd02ea701668f7d45/src/AutoConnectFS.h#L60

    Change to:
    ```cpp
    #define AUTOCONNECT_APPLIED_FILECLASS     fs::LITTLEFSFS
    ```

    ---
    https://github.com/Hieromon/PageBuilder/blob/944b9535addc5d5a24a0d124687a4ddfb15e2d35/src/PageBuilder.h#L93

    Change to:
    ```cpp
    #define PB_APPLIED_FILESYSTEM             LITTLEFS
    ```

    ---
    https://github.com/Hieromon/PageBuilder/blob/944b9535addc5d5a24a0d124687a4ddfb15e2d35/src/PageBuilder.h#L95

    Change to:
    ```cpp
    #define PB_APPLIED_FILECLASS              fs::LITTLEFSFS
    ```

2. Modify the [`LITTLEFS.h`](https://github.com/lorol/LITTLEFS/blob/master/src/LITTLEFS.h), [`LITTLEFS.cpp`](https://github.com/lorol/LITTLEFS/blob/master/src/LITTLEFS.cpp) in [*LittleFS_esp32*](https://github.com/lorol/LITTLEFS) library source

    If you apply this fix, you will no longer be able to build your legacy sketches with the *LittleFS_esp32* library applied. This will cause a compile error.

    ---
    https://github.com/lorol/LITTLEFS/blob/f0817ca5264745acce697092de2bf218b3aa5b2e/src/LITTLEFS.h#L22

    Change to:
    ```cpp
    class LittleFSFS : public FS
    ```

    ---
    https://github.com/lorol/LITTLEFS/blob/f0817ca5264745acce697092de2bf218b3aa5b2e/src/LITTLEFS.h#L25-L26

    Change to:
    ```cpp
    LittleFSFS();
    ~LittleFSFS();
    ```

    ---
    https://github.com/lorol/LITTLEFS/blob/f0817ca5264745acce697092de2bf218b3aa5b2e/src/LITTLEFS.h#L39

    Change to:
    ```cpp
	extern fs::LittleFSFS LittleFS;
    ```

    ---
    https://github.com/lorol/LITTLEFS/blob/f0817ca5264745acce697092de2bf218b3aa5b2e/src/LITTLEFS.cpp#L48

    Change to:
    ```cpp
    LittleFSFS::LittleFSFS() : FS(FSImplPtr(new LITTLEFSImpl())), partitionLabel_(NULL)
    ```

    ---
    https://github.com/lorol/LITTLEFS/blob/f0817ca5264745acce697092de2bf218b3aa5b2e/src/LITTLEFS.cpp#L52

    Change to:
    ```cpp
    LittleFSFS::~LittleFSFS()
    ```

    ---
    https://github.com/lorol/LITTLEFS/blob/f0817ca5264745acce697092de2bf218b3aa5b2e/src/LITTLEFS.cpp#L97

    Change to:
    ```cpp
    void LittleFSFS::end()
    ```

    ---
    https://github.com/lorol/LITTLEFS/blob/f0817ca5264745acce697092de2bf218b3aa5b2e/src/LITTLEFS.cpp#L109

    Change to:
    ```cpp
    bool LittleFSFS::format()
    ```

    ---
    https://github.com/lorol/LITTLEFS/blob/f0817ca5264745acce697092de2bf218b3aa5b2e/src/LITTLEFS.cpp#L121

    Change to:
    ```cpp
    size_t LittleFSFS::totalBytes()
    ```

    ---
    https://github.com/lorol/LITTLEFS/blob/f0817ca5264745acce697092de2bf218b3aa5b2e/src/LITTLEFS.cpp#L130

    Change to:
    ```cpp
    size_t LittleFSFS::usedBytes()
    ```

    ---
    https://github.com/lorol/LITTLEFS/blob/f0817ca5264745acce697092de2bf218b3aa5b2e/src/LITTLEFS.cpp#L139

    Change to:
    ```cpp
    LittleFSFS LittleFS;
    ```

**Comment by HankLloydRight** [comment]
Wow, thank you Hieromon!  You are the best.  Thanks for the extremely detailed explanation. 

Since I don't use the Arduino IDE and only use Platform.io, it seems that Option #1 is the way to go. (modify AutoConnectFS.h and PageBuilder.h)  I made those changes and now the code compiles cleanly and my code size went down about 29k. No change in free stack space though.

Interestingly, the SPIFFS library is still in the dependency tree. It's remained unchanged from what I posted above.  Is that expected?

**Reply by Hieromon** [reply]
@HankLloydRight
> the SPIFFS library is still in the dependency tree.

In the past, I have also found spiffs in the dependency tree output by PIO and wondered about it. But that seems to be the specification of the PIO [Library Dependency Finder (LDF)](https://docs.platformio.org/en/stable/librarymanager/ldf.html#library-dependency-finder-ldf). If `lib_ldf_mode = deep` specified, LDF will detect `#include` directives without evaluating the preprocessor conditional expressions. The following quotation is from the [PIO documentation](https://docs.platformio.org/en/stable/librarymanager/ldf.html#dependency-finder-mode) regarding the `deep` option.

> deep:
Parses ALL C/C++ source files of the project and parses ALL C/C++ source files of the each found dependency (recursively). **Does not evaluate** [C/C++ Preprocessor conditional syntax](https://docs.platformio.org/en/stable/librarymanager/ldf.html#ldf-c-cond-syntax).

And also, PIO determines what to compile based on its own detection, so `SPIFFS.cpp` is compiled. However, there is no text in the compiled object that refers to the SPIFFS class, so the linker does not bind it. You can check it using the [Inspector](https://docs.platformio.org/en/latest/home/index.html#project-inspect) of the PlatformIO IDE. 

After building the project, start the Inspector, open [Symbols](https://docs.platformio.org/en/latest/home/index.html#firmware-symbols) search and search for **spiffs**. You will see that the firmware does not contain any spiffs. And you do the same search using **littlefs** as a keyword, you will see that littlefs is present in some of the text sections.

**Reply by HankLloydRight** [reply]
Hieromon,
 Again, thank you for all you do!  

I turned off `lib_ldf_mode = deep` and still the same result. 

But you're right, in the inspector I didn't find any references to SPIFF when I run the Memory Inspector/Symbols.
The 'Code Inspector' crashes out with pages and pages of compiler errors, so I have no idea how to get that to work.

Do you take any contributions for your work or a preferred charity for donations?  
Your library and dedication to helping individual users is just so awesome, how can I support your efforts?
 Thanks.

**Reply by Hieromon** [reply]
@HankLloydRight Thank you for appreciating my work. Fortunately, at the present time, I am earning enough to stay healthy with my family through my regular job. Therefore, for the time being, I am able to continue to maintain and enhance the library without any donations. Of course, tomorrow never knows.

What I expect to most users is the sharing of knowledge.

The AutoConnect source code is not so elegant. Also, AutoConnectAux's method of accessing AutoConnectElements does not fully utilize the capabilities of C++. This is largely due to the limitations of statically-typed languages, which must be able to be deployed in the limited memory space of the ESP32 and ESP8266, which can be difficult for users who primarily use Arduino as their programming framework.
There are engineers all over the world with the ability to solve these challenges. I believe that if I can get advice from them, AutoConnect will become a more user-friendly library for other users.

I would also like to hear from general Arduino users to provide use cases they are experiencing. It will bring me new discoveries and a strong motivation to develop even better libraries. I expect you to provide it to me.

You have taught me the following through this topic:

"AutoConnect needs to support the legacy LittleFS_esp32 in ESP32 core 1.0.6 with PIO."

Without your contribution, I would not have supported the legacy LittleFS_esp32 library in ESP32 core 1.0.6. Its discussion in my mind was already over. However, your post may have saved other users. It's only a possibility, though. But I find that very valuable. 

Thank you for your contribution.

**Reply by HankLloydRight** [reply]
> I would also like to hear from general Arduino users to provide use cases they are experiencing. It will bring me new discoveries and a strong motivation to develop even better libraries. I expect you to provide it to me.

Just wanted to show you what I'm working on using your libraries:  https://ildac32.com/ 
And of course AutoConnect is included on my list of credits: https://ildac32.com/credits.php
You can see the AutoConnect library in action at the 3:00 mark in this video:  https://youtu.be/kLh59MJFXyc

thanks!

**Reply by Hieromon** [reply]
Awesome!, Thank you inform to me. In addition to being capable as a standalone ILDA player, it can be controlled interactively using a cell phone! I'm happy to contribute to the high degree project like you're making.

By the way, observing the video you introduced, I assume that the ILDAC-32 will encounter a situation where it loses WiFi connection during WiFi control.  Do you use [AutoConnectConfig::autoReconnect with reconnectInterval](https://hieromon.github.io/AutoConnect/adconnection.html#automatic-reconnect-background)? Does it work as expected for you?

**Comment by Hieromon** [comment]
Support for the [LittleFS_esp32](https://github.com/lorol/LITTLEFS) legacy library in ESP32 core 1.0.6 has been released as [AutoConnect v1.3.4](https://github.com/Hieromon/AutoConnect/releases/tag/v1.3.4) and [PageBuilder 1.5.3](https://github.com/Hieromon/PageBuilder/releases/tag/v1.5.3).

**Comment by Hieromon** [comment]
@HankLloydRight Hello, Today I have something to let you know.
PlatformIO has released a development platform as [espressif32](https://github.com/platformio/platform-espressif32/releases/tag/v4.0.0) equivalent to the ESP32 Arduino Core 2.0.0. It is the ESP-IDF4.4 toolchain-based environment and includes LittleFS ported for ESP32. Currently, its latest head has progressed to [4.1.0](https://github.com/platformio/platform-espressif32/releases/tag/v4.1.0). This allows us to forego the [legacy version of the LittleFS](https://github.com/lorol/LITTLEFS) library for ESP32.

The platform-espressif32 platform maintains compatibility with ESP32 Arduino Core 2.0. So the LittleFS_esp32 library installed individually in the Arduino library folder is no longer needed. And if it is left behind, PlatformIO's LDF will detect the LittleFS_esp32, which is supposed to be a residue. Upon that detection, the FS class is expanded to LITTLEFS. In other words, the FS class will be expanded to LITTLEFS prior to 2.0.0 (which is legacy), even though it is a library dependency for Arduino Core 2.0.0 or later, resulting in a compile error for AutoConnect. This error can be resolved by simply deleting the LittleFS_esp32 folder from the Arduino library folder.

If you are already aware of this, then my concern is unfounded. Thank you.

**Reply by HankLloydRight** [reply]
Hello.. Thank you. I was not aware of the new espressif32 4.2.0 platform.

I upgraded my project to espressif32 4.2.0 and Autoconnect 1.3.4 and changed a few lines of code and it compiled and ran.
I had to tweak a few lines of code when pulling files from the LittleFS partition as there seems to be some differences now with the embedded version compared with the legacy version, but I got it to work.

But the bigger problem I have now (unrelated to Autoconnect) is that my interrupt loop and CPU pinned loops are totally messed up. Before the upgrade using espressif32 @ 3.5.0, I had CPU1 handling all the user interface I/O, SD card I/O, Wifi, and BLE I/O while CPU0 was *only* sending data points to the DACs at 30khz-70khz.  I am using an interrupt timer by setting a timerBegin with timerAttachInterrupt, timerAlarmWrite, and timerAlarmEnable to fire the function to send data to the DAC, and it was working perfectly and an extremely stable interrupt clock.

But once I upgraded to 4.2.0, the loop timing is very unstable and now crashes when I go to change the interrupt timers.  Something between 3.5.0 and 4.2.0 totally changed how internal tasks are assigned to CPU0 and CPU1, and it totally wrecked my stable running project.

I just downgraded to espressif32@3.5.0 and back to the legacy LITTLEFS (and Autoconnect 1.3.4) and all is restored back to the previous state with my project. But I have no idea how to go about debugging the new framework to solve these problems.   I think I'll have to pin future development to espressif32@3.5.0.

**Reply by Hieromon** [reply]
Hello @HankLloydRight, I was a little curious about your mention of the poor performance of the peripheral, so I did some research. I have not yet been able to fully identify the factors, but I will tell you that I have found something that may be relevant to this topic.
The PlatformIO 4.2.0 framework is based on esp32-arduino core 2.0.2, which has significant differences in peripheral access compared to arduino core 1.0.6. In arduino core 1.0.6, most peripheral accesses can read and write directly to registers, but in core 2.0.2, they have been refactored to use IDF. Thus, everything is now IRQ-driven, and CPU overhead has steadily increased.
This constraint comes to the fore as a mysterious behavior in applications that assumes interrupt handlers to complete within a certain amount time. At first sight, it may be difficult to find the cause of the problem. Especially if the application drives many peripherals like your project, I think that the cooperation between the peripherals will be completely cut off.

In other words, are the peripheral libraries that your project depends on really thread-safe? That is, as long as the CPU is running in a pseudo task slice, it doesn't matter. Direct reading of registers is the most extreme of that. Essentially, the peripheral read process should be separate from the writing process. Before processing a read, we need to confirm the commit of the write transaction (usually surrounded by beginTransaction and endTransaction). As a result of the adoption of ESP-IDF 4.4 in Arduino Core 2.0.2, thread-safe and asynchronization transactions are required assuming a complete IRQ drive. I've seen many Arduino libraries with their own wild logic that doesn't follow that approach. (Even that are well-known and frequently used) And I have observed that many of the libraries distributed with the esp32-arduino core 2.0.2 and later have been rewritten to be thread-safe using semaphores and MUTEX. 

If the loop timing instability you are encountering is the result of access to the periphery being refactored to an IRQ-driven model, then the timer interrupt technique will not solve the problem. It would require transaction-based asynchronization. This could force a significant change in the structure of your project, but the only benefit you would gain from it would probably be a long-term service of IDF 4.4. (Espressif has already ended support for the IDF 3.x series)
I guess in a way it might be wise to pin your project to the arduino core 1.0.6 (a combination of IDF-3.x and espressif32@3.5.0).

**Reply by HankLloydRight** [reply]
@hieromon -- Your knowledge, insight, and willingness to help people is phenomenal.  I can't thank you enough. 

What you described is a little beyond my knowledge base, but I think it totally hits the nail on the head. Let me elaborate:

 In my project, I'm using 4 12-bit SPI DACs (MCP4922) to send analog data to the laser projector. Bit banging in this case is actually much faster then hardware SPI.  

When I started out, I was using this github library https://github.com/RobTillaart/MCP_DAC to communicate with the DACs. But his library is for generic Arduino projects and uses analogWrite(), which is very slow and did not provide the performance I wanted, but his code is portable. 

So I took his code and changed it to use direct register writes on the ESP32 like this:

Initialization code:
` REG_WRITE ( GPIO_ENABLE_REG, BIT23+BIT5+BIT16+BIT27+BIT18+BIT17+BIT0 ) ;`

And then these functions to write to the DACs on those pins using the pin registers:

```
void MCP_DAC::fastWriteAB_direct(uint16_t v1,uint16_t v2) {
    REG_WRITE ( GPIO_OUT_W1TC_REG, __select);
    swSPI_transfer_directio((uint8_t)((0x3000 | v1) >> 8));
    swSPI_transfer_directio((uint8_t)(v1 & 0xFF));
    REG_WRITE ( GPIO_OUT_W1TS_REG, __select);
    REG_WRITE ( GPIO_OUT_W1TC_REG, __select);
    swSPI_transfer_directio((uint8_t)((0xB000 | v2) >> 8));
    swSPI_transfer_directio((uint8_t)(v2 & 0xFF));
    REG_WRITE ( GPIO_OUT_W1TS_REG, __select);
}

void MCP_DAC::swSPI_transfer_directio(uint8_t val) {
  for (uint8_t mask = 0x80; mask; mask >>= 1)  {
    if (val & mask) REG_WRITE ( GPIO_OUT_W1TS_REG, __mosi ) ;
    else REG_WRITE ( GPIO_OUT_W1TC_REG, __mosi ) ;
    REG_WRITE ( GPIO_OUT_W1TS_REG, __clock);
    REG_WRITE ( GPIO_OUT_W1TC_REG, __clock);
  }
}
```

This gave me the performance I needed.  I then setup an interrupt timer to send points to the DACs like this:

```
timer = timerBegin(0, prescaler , true);
timerAttachInterrupt(timer, &onTick, true);
timerAlarmWrite(timer, 1, true);
timerAlarmEnable(timer);
void IRAM_ATTR onTick() {
	//portENTER_CRITICAL_ISR(&timerMux);
	interrupts++;
	//portEXIT_CRITICAL_ISR(&timerMux);
}
```

The portENTER/EXIT_CRITICAL_ISR functions are commented out because that slowed down the performance too much, and leaving them out had no negative effects.  Maybe that's my problem!  


Then in the main loop(){} function I have:
```
	if (interrupts>0) { // when there is a clock tick, do something
                SendPoints(); // send points to DACs from memory
		//portENTER_CRITICAL(&timerMux);
		interrupts--;
		//portEXIT_CRITICAL(&timerMux);
	}
```

That's the most important parts of my code.  When I compile this for espressif32 >= 4.0, the timing is way off and the patterns drawn are very unstable as you described.  As such, I've already pinned my code to espressif32@3.5.0.   And from the sound of it, it looks like I won't ever be able to move past that.   But at least now the LittleFS integration with Autocorrect is working great!  

Thank you again for your help and insight.
